proxy/InkAPIInternal.h

```diff
@@ -117,6 +117,12 @@ class APIHook
   APIHook *next() const;
   APIHook *prev() const;
   LINK(APIHook, m_link);
+
+  // This is like invoke(), but allows for blocking on continuation mutexes.  It is a hack, calling it can block
+  // the calling thread.  It should be replaced with a version with a continuation parameter, providing a continuation
+  // to be called when the hook handling is complete.
+  //
+  int blocking_invoke(int event, void *edata) const;
```

Comment:
The APIHook::invoke() (what was originally used instead of this) will, if the mutex of the continuation is already locked, do an ink_release_assert().  So that effectively means you can't have a continuation mutex.  I needed a mutex on a continuation in an internal Yahoo plugin.  APIHook::blocking_invoke()  will just block on the mutex.  If there is time someday, all hooks should be made to work like the global and transaction hooks.  Try the mutex, and if you can't lock it, immediate reschedule the continuation.  That's the sense in which this is a hack.